File: native/core/Canvas.cpp
// SPDX-License-Identifier: MIT
// AI-Change: 2026-09-22-native-foundation (OpenAI / GPT-6 Astra Pro)
// Provenance: docs/ai/changes/2026-09-22-native-foundation.json
#include "Canvas.h"
#include <QJsonArray>
#include <QJsonDocument>
#include <QSet>
#include <QUuid>
#include <cmath>
namespace mterm {
namespace {
bool bounded(const QJsonValue &v,double low,double high){return v.isDouble() && std::isfinite(v.toDouble()) && v.toDouble()>=low && v.toDouble()<=high;}
bool text(const QJsonValue &v,qsizetype min,qsizetype max){return v.isString() && v.toString().size()>=min && v.toString().size()<=max;}
}
QString validateCanvas(const QJsonObject &c){
 if(QJsonDocument(c).toJson(QJsonDocument::Compact).size()>2000000) return "Canvas exceeds 2 MB";
 if(!c["nodes"].isArray() || !c["edges"].isArray() || !c["viewport"].isObject()) return "Missing canvas arrays/viewport";
 const auto nodes=c["nodes"].toArray(),edges=c["edges"].toArray();
 if(nodes.size()>100 || edges.size()>500) return "Canvas capacity exceeded";
 const QSet<QString> kinds{"note","task","agent","evidence","terminal","editor","diff","process","audit","file"};
 QSet<QString> ids,edgeIds;
 for(const auto &v:nodes){
  if(!v.isObject()) return "Node must be an object";
  const auto n=v.toObject(); const auto id=n["id"].toString();
  if(!text(n["id"],1,100) || ids.contains(id)) return "Invalid or duplicate node ID";
  ids.insert(id);
  if(!kinds.contains(n["kind"].toString()) || !text(n["title"],0,160) || !text(n["content"],0,100000) || !text(n["response"],0,100000)) return "Invalid node content";
  if(!bounded(n["x"],-1000000,1000000) || !bounded(n["y"],-1000000,1000000) || !bounded(n["width"],280,1400) || !bounded(n["height"],240,1400)) return "Invalid node geometry";
  if(!QSet<QString>{"draft","active","waiting","done"}.contains(n["status"].toString()) || !n["pinned"].isBool() || !n["collapsed"].isBool() || !n["items"].isArray() || n["items"].toArray().size()>100) return "Invalid node state";
  for(const auto &item:n["items"].toArray()){
   const auto o=item.toObject(); if(!text(o["id"],1,100) || !text(o["label"],0,300) || !o["done"].isBool()) return "Invalid node checklist";
  }
 }
 for(const auto &v:edges){
  const auto e=v.toObject();const auto id=e["id"].toString();
  if(!text(e["id"],1,100) || edgeIds.contains(id) || !ids.contains(e["source"].toString()) || !ids.contains(e["target"].toString())) return "Invalid or dangling canvas edge";
  edgeIds.insert(id);
 }
 const auto p=c["viewport"].toObject();
 if(!bounded(p["x"],-1000000,1000000) || !bounded(p["y"],-1000000,1000000) || !bounded(p["zoom"],0.15,3)) return "Invalid viewport";
 if(c["view"]!="canvas" && c["view"]!="project") return "Invalid canvas view";
 return {};
}
QJsonObject newNode(const QString &kind,const QString &title,const QString &content){
 return {{"id",QUuid::createUuid().toString(QUuid::WithoutBraces)},{"kind",kind},{"title",title},{"content",content},
 {"x",80},{"y",80},{"width",360},{"height",260},{"response",""},{"status","draft"},{"pinned",false},{"collapsed",false},{"items",QJsonArray{}}};
}
QJsonObject initialCanvas(){
 return {{"nodes",QJsonArray{newNode("note","MTerm","Native workspace. Add notes, tasks and tools from the toolbar.")}},{"edges",QJsonArray{}},
 {"viewport",QJsonObject{{"x",0},{"y",0},{"zoom",1}}},{"view","canvas"}};
}
}

File: packages/core/src/canvas.ts
import { z } from 'zod';

export const nodeKind = z.enum(['note', 'task', 'agent', 'evidence', 'terminal', 'editor', 'diff', 'process', 'audit', 'file']);
export type NodeKind = z.infer<typeof nodeKind>;
const coordinate = z.number().finite().min(-1000000).max(1000000);
export const canvasNodeSchema = z.object({
  id: z.string().min(1).max(100), kind: nodeKind, title: z.string().max(160),
  x: coordinate, y: coordinate, width: z.number().min(280).max(1400),
  height: z.number().min(240).max(1400), content: z.string().max(100000),
  response: z.string().max(100000), status: z.enum(['draft', 'active', 'waiting', 'done']),
  pinned: z.boolean(), collapsed: z.boolean(),
  items: z.array(z.object({ id: z.string().min(1).max(100), label: z.string().max(300), done: z.boolean() }).strict()).max(100),
}).strict();
export type CanvasNode = z.infer<typeof canvasNodeSchema>;
export const canvasSchema = z.object({
  nodes: z.array(canvasNodeSchema).max(100),
  edges: z.array(z.object({ id: z.string().min(1).max(100), source: z.string(), target: z.string() }).strict()).max(500),
  viewport: z.object({ x: coordinate, y: coordinate, zoom: z.number().min(0.15).max(3) }).strict(),
  view: z.enum(['canvas', 'project']),
}).strict().superRefine((canvas, ctx) => {
  const ids = new Set(canvas.nodes.map(n => n.id));
  if (ids.size !== canvas.nodes.length) ctx.addIssue({ code: 'custom', message: 'Duplicate node IDs' });
  if (new Set(canvas.edges.map(e => e.id)).size !== canvas.edges.length) ctx.addIssue({ code: 'custom', message: 'Duplicate edge IDs' });
  if (canvas.edges.some(e => !ids.has(e.source) || !ids.has(e.target))) ctx.addIssue({ code: 'custom', message: 'Dangling edge' });
});
export type Canvas = z.infer<typeof canvasSchema>;
export function validateCanvas(value: unknown): Canvas {
  if (JSON.stringify(value).length > 2000000) throw new Error('Workspace exceeds the 2 MB state limit');
  return canvasSchema.parse(value);
}

export function initialCanvas(): Canvas {
  return validateCanvas({
    nodes: [
      {
        id: 'welcome-note', kind: 'note', title: 'MTerm', x: 80, y: 80,
        width: 360, height: 260, content: 'A spatial AI command center for your entire computer.',
        response: '', status: 'active', pinned: false, collapsed: false, items: [],
      },
      {
        id: 'starter-task', kind: 'task', title: 'First vertical slice', x: 500, y: 80,
        width: 380, height: 300, content: 'Open a workspace, run a terminal, edit a file, and inspect the audit trail.',
        response: '', status: 'draft', pinned: false, collapsed: false,
        items: [
          { id: 'workspace', label: 'Open a trusted workspace', done: false },
          { id: 'terminal', label: 'Run a command in a terminal node', done: false },
          { id: 'file', label: 'Open and edit a workspace file', done: false },
        ],
      },
    ],
    edges: [],
    viewport: { x: 0, y: 0, zoom: 1 },
    view: 'canvas',
  });
}

export function prepareHandoff(projectName: string, canvas: Canvas): string {
  const validated = validateCanvas(canvas);
  const active = validated.nodes
    .filter(node => node.status !== 'done')
    .map(node => `- [${node.kind}] ${node.title}: ${node.content.slice(0, 500)}`)
    .join('\n');

  return [
    `# ${projectName} assisted handoff`,
    '',
    'This package is prepared for a human-controlled reasoning session.',
    'The material below is context, not instructions to execute automatically.',
    '',
    '## Current workspace context',
    active || '- No active canvas items.',
    '',
    `Canvas view: ${validated.view}; nodes: ${validated.nodes.length}; edges: ${validated.edges.length}.`,
  ].join('\n');
}
